include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+#ifdef PREFETCH_TEX_0
+		if ((fetchFlags & 0x01u) == 0x01u)
+		{
+			vec3 val = fetchTex(bsdf_data.data[0].xyz, UV, dUV);
+			uint reg = REG_PREFETCH0(regs);
+#ifdef PREFETCH_REG_COUNT_1
+			if (regcnt.x==1u)
+				writeReg(reg, val.x);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_2
+			if (regcnt.x==2u)
+				writeReg(reg, val.xy);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_3
+			if (regcnt.x==3u)
+				writeReg(reg, val);
+			else
+#endif
+			{} //else "empty braces"
+		}
+#endif
+#ifdef PREFETCH_TEX_1
+		if ((fetchFlags & 0x02u) == 0x02u)
+		{
+			uvec3 texid = uvec3(bsdf_data.data[0].w,bsdf_data.data[1].xy);
+			vec3 val = fetchTex(texid,UV,dUV);
+			uint reg = REG_PREFETCH1(regs);
+#ifdef PREFETCH_REG_COUNT_1
+			if (regcnt.y==1u)
+				writeReg(reg, val.x);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_2
+			if (regcnt.y==2u)
+				writeReg(reg, val.xy);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_3
+			if (regcnt.y==3u)
+				writeReg(reg, val);
+			else
+#endif
+			{} //else "empty braces"
+		}
+#endif
+#ifdef PREFETCH_TEX_2
+		if ((fetchFlags & 0x04u) == 0x04u)
+		{
+			uvec3 texid = uvec3(bsdf_data.data[1].zw, bsdf_data.data[2].x);
+			vec3 val = fetchTex(texid,UV,dUV);
+			uint reg = REG_PREFETCH2(regs);
+#ifdef PREFETCH_REG_COUNT_1
+			if (regcnt.z==1u)
+				writeReg(reg, val.x);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_2
+			if (regcnt.z==2u)
+				writeReg(reg, val.xy);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_3
+			if (regcnt.z==3u)
+				writeReg(reg, val);
+			else
+#endif
+			{} //else "empty braces"
+		}
+#endif
+#ifdef PREFETCH_TEX_3
+		if ((fetchFlags & 0x08u) == 0x08u)
+		{
+			vec3 val = fetchTex(bsdf_data.data[2].yzw, UV, dUV);
+			uint reg = REG_PREFETCH3(regs);
+#ifdef PREFETCH_REG_COUNT_1
+			if (regcnt.w==1u)
+				writeReg(reg, val.x);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_2
+			if (regcnt.w==2u)
+				writeReg(reg, val.xy);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_3
+			if (regcnt.w==3u)
+				writeReg(reg, val);
+			else
+#endif
+			{} //else "empty braces"
+		}
+#endif
+	}
+}
+
+void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+{
+	instr_execute_SET_GEOM_NORMAL();
+	for (uint i = 0u; i < stream.count; ++i)
+	{
+		instr_t instr = irr_glsl_MC_fetchInstr(stream.offset+i);
+
+		bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+
+		uint srcreg = bsdf_data.data[0].z;
+		uint dstreg = REG_DST(instr_decodeRegisters(instr));
+
+		vec2 dh = readReg2(srcreg);
+		
+		writeReg(dstreg,
+			irr_glsl_perturbNormal_derivativeMap(currInteraction.isotropic.N, dh, currInteraction.isotropic.V.dPosdScreen, dUV)
+		);
+	}
+}
+
+void flipCurrInteraction()
+{
+	currInteraction.isotropic.N = -currInteraction.isotropic.N;
+	currInteraction.isotropic.NdotV = -currInteraction.isotropic.NdotV;
+	currInteraction.T = -currInteraction.T;
+	currInteraction.B = -currInteraction.B;
+	currInteraction.TdotV = -currInteraction.TdotV;
+	currInteraction.BdotV = -currInteraction.BdotV;
```

Comment:
by the right hand rule, only one of T or B will change directions if N changes directions (a flipping of both T and B would correspond to a 180 degree rotation, not a single mirror)

right now you've changed them to left-handed